### PR TITLE
deprecation.pod

### PR DESCRIPTION
**Pull request by richsalz**
I could not understand deprecation.pod  I rewrote it. I hope this is more clear.

### CODE REVIEW

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ }
+ #endif
+
+If a deprecated function is needed elsewhere within OpenSSL, do not
+wrap the implementation because some build configurations will "ifdef out"
+the implementation. Instead, a file that implements publically-deprecated
+function should include this near the beginning of its translation unit:
+
+ /*
+  * RSA low level APIs are deprecated for public use, but are kept for
```
Comment (paulidale) [inline doc/internal/man7/OSSL_DEPRECATED.pod:124]: For consistency, I'd suggest HMAC instead of RSA.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  */
+ #include "internal/deprecated.h"
+
+
+Any OpenSSL source file that needs to use a deprecated function, should
+add this line (or equivalent) at the start of the translation unit,
+perhaps with an explanatory comment:
+
+ /*
+  * Still using legacy RSA_ functions.
```
Comment (paulidale) [inline doc/internal/man7/OSSL_DEPRECATED.pod:135]: HMAC again?

**Review by paulidale — APPROVED** [review]
A big improvement.

Consistency suggestions only.  Approved either way.

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 1 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod]: really good point, done!

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 2 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod]: yes.  I also changed the comment wording.

**Comment by richsalz** [comment]
Rebased, squashed, added a fixup comment.  thanks for the quick review.

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
So your variant is a wall of text.  As a quick reference when you just want a reminder if you should use `#include "internal/deprecated.h"` or `#define OPENSSL_SUPPRESS_DEPRECATED`, I find this not working, there are no quick visual cues.

Furthermore, the explanation for each of those two variants leaves me confused.  On further scrutiny, they seem not quite correct.  They are to be used *both* where deprecated symbols are used and where they are implemented.  The big difference is that: 

- `#include "internal/deprecated.h"` is used for translation units that use or implement symbols that we plan to keep internally after it's been removed from public headers.  Neither the use of the implementation should be guarded with `#ifndef OPENSSL_NO_DEPRECATED_version`
- `#define OPENSSL_SUPPRESS_DEPRECATED` is used for translation units that use or implement symbols that we plan to remove.  Both the use and the implementation needs to be guarded with `#ifndef OPENSSL_NO_DEPRECATED_version`.

Your text tells a different story.  I do realise, however, that my text isn't as clear as I wish on the matter, but that's simple to correct.

**Comment by richsalz** [comment]
What I wrote might be a "subset" of what you describe, but it seems to work and follows existing practice and is less invasive.

**Comment by levitte** [comment]
> What I wrote might be a "subset" of what you describe, but it seems to work and follows existing practice and is less invasive.

"less invasive"?????  I have no idea what you mean with that.  In terms of code, the result should be the same...

**Comment by richsalz** [comment]
I thought it meant less changes to deprecate things; for example, not wrapping everything in in ```ifndef``` but maybe I just did not understand the existing manpage.

**Comment by levitte** [comment]
The only thing extra with the new deprecation way is the need to guard the deprecated stuff with `OPENSSL_NO_DEPRECATED_version` in the public headers, and you've copied the example for that verbatim from `deprecation.pod`.

The rest, i.e. having the same guards in the source code, and need to define `OPENSSL_SUPPRESS_DEPRECATED` or including `"internal/deprecated.h"` is already existing and required stuff and has been so for a while.

**Comment by richsalz** [comment]
I included the rationale for why (and why not) to do things, I removed needless scenario's, and think mine is more clear. I did it because I was working on #13171 and I didn't understand the existing document.

I offer this to the project, which can accept it or not, neither way will hurt my feelings.

**Comment by levitte** [comment]
Unfortunately, this contains too much confusion and incorrectness.

-1

**Comment by paulidale** [comment]
Assuming the -1 is OTC, I'd added the tag.

**Comment by levitte** [comment]
Not sure how it's OTC.  I simply disapprove.  But if that makes it worth an OTC discussion, then sure.

**Comment by paulidale** [comment]
Clearing the hold since it wasn't intended.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+warning that HMAC_Init() is deprecated.
+
+To make it possible to hide the declaration completely, all
+deprecations should be wrapped inside an ifdef. For compatibility with
+other OpenSSL macros, a double negative must be used:
+
+ #ifndef OPENSSL_NO_DEPRECATED_1_1_0
+ OSSL_DEPRECATEDIN_1_1_0 __owur int
+ HMAC_Init(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md);
+ #endif
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:95]: I would like a visual cue, such as a visible section title, that leads the eye here when you want to remember how it should be done in a public header file.  (`deprecated.pod` has that)

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+C<typedef>, or a variable declaration.
+When deprecating items in a header file, it might make sense to move
+deprecations to one place, rather than putting multiple instances of
+pre-processor C<ifndef> lines throughout the file.
+
+To remove the actual implementation, use the wrapper:
+
+ #ifndef OPENSSL_NO_DEPRECATED_1_1_0
+ OSSL_DEPRECATEDIN_1_1_0 __owur int
+ HMAC_Init(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md)
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:113]: I would like a visual cue, such as a visible section title, that leads the eye here when you want to needs to be done with an implementation that we plan to remove entirely.  (`deprecated.pod` has that)

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+Hiding the declaration completely is done when OpenSSL is configured to
+remove old API's, via the C<--api> configuration parameter.
+
+The wrapper must be used when deprecating a C<define>,
+C<typedef>, or a variable declaration.
+When deprecating items in a header file, it might make sense to move
+deprecations to one place, rather than putting multiple instances of
+pre-processor C<ifndef> lines throughout the file.
+
+To remove the actual implementation, use the wrapper:
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:109]: Precisely speaking, this is not actual removal.  Actual removal is actually removing the implementation, i.e. hit the delete-line button until it's actually *gone* from the source.  That is an action that we will perform some day.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+ #endif
+
+Note that the script that parses the header files is not a full C
+parser, so the C<ifndef> must be on a line by itself.
+
+Hiding the declaration completely is done when OpenSSL is configured to
+remove old API's, via the C<--api> configuration parameter.
+
+The wrapper must be used when deprecating a C<define>,
+C<typedef>, or a variable declaration.
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:104]: Also with a function declaration.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ HMAC_Init(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md)
+ {
+   ...
+   return ret;
+ }
+ #endif
+
+If a deprecated function is needed elsewhere within OpenSSL, do not
+wrap the implementation because some build configurations will "ifdef out"
+the implementation. Instead, a file that implements publically-deprecated
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:122]: Not just implements, but also the source files that *use* publicly deprecated functions.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+function should include this near the beginning of its translation unit:
+
+ /*
+  * HMAC low level APIs are deprecated for public use, but are kept for
+  * internal use.
+  */
+ #include "internal/deprecated.h"
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:129]: I would like a visual cue, such as a visible section title, that leads the eye here when you want to needs to be done using a function that's publicly deprecated but will be kept internally.  (`deprecated.pod` has that)

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 7 above)
Comment (levitte) [reply doc/internal/man7/OSSL_DEPRECATED.pod:122]: Also, this paragraph is confusing.  Like I said, the use of ENGINE stuff is still currently needed within OpenSSL, but ENGINE is deprecate with the intent to drop it entirely, and I can't see from this text if I should use `#include "internal/deprecated.h"` or `#define OPENSSL_SUPPRESS_DEPRECATED` in files where ENGINE stuff is used.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+
+Any OpenSSL source file that needs to use a deprecated function, should
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:132]: `#define OPENSSL_SUPPRESS_DEPRECATED` is for any file that *implements* or *uses* anything that's deprecated with the intent to actually remove it in the future.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+ /*
+  * This file uses the publically-deprecated HMAC functions.
+  */
+ #define OPENSSL_SUPPRESS_DEPRECATED
+
+In a subsequent release, the declaration should be moved to an internal
+header file and the wrapper and deprecation attribute should be removed.
+A more complete solution is to rewrite the code that uses the deprecated
+API's.
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:144]: This paragraph looks related to `#define OPENSSL_SUPPRESS_DEPRECATED`, but at least the first sentence is actually related to `#include "internal/deprecated.h"`...

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+
+Any OpenSSL source file that needs to use a deprecated function, should
+add this line (or equivalent) at the start of the translation unit,
+perhaps with an explanatory comment:
+
+ /*
+  * This file uses the publically-deprecated HMAC functions.
+  */
+ #define OPENSSL_SUPPRESS_DEPRECATED
```
Comment (levitte) [inline doc/internal/man7/OSSL_DEPRECATED.pod:139]: I would like a visual cue, such as a visible section title, that leads the eye here when you want to needs to be done when using a function that's publicly deprecated and that will be removed.  (`deprecated.pod` has that)

**Review by levitte — COMMENTED** [review]
I detect a lot of confusion on the matter of deprecation, and especially what's going to happen with the deprecated declarations, as well as with the implementations.  I wonder what I can do to clear that up...

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+ OSSL_DEPRECATEDIN_0_9_8
+ OSSL_DEPRECATEDIN_0_9_8_FOR(msg)
+
+=head1 DESCRIPTION
+
+Deprecation means to indicate that a symbol (usually a function, but can also
+be a define and less commonly a type or even variable) should not be used
+any more.
+OpenSSL supports multiple types of deprecation, including compile-time
+warnings, all the way to the removal of the implementation.
```
[the remark this answers is not in the document]
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:57]: I rewrote the first couple of paragraphs to make the separation between deprecation and removal (via ifdef and the source) more clear.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+This is controlled by two configurations parameters: C<--api=> to
+specify a version, and C<no-deprecated> to control actual removal.
+These are documented in the F<INSTALL.md> file provided with the source.
+As a third case, sometimes an item is deprecated, but it is still used
+within the OpenSSL software itself.
```
[the remark this answers is not in the document]
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:62]: This is rewritten.

File: doc/internal/man7/OSSL_DEPRECATED.pod [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+ #ifndef OPENSSL_NO_DEPRECATED_1_1_0
+ OSSL_DEPRECATEDIN_1_1_0 __owur int
+ HMAC_Init(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md);
+ #endif
+
+Note that the script that parses the header files is not a full C
+parser, so the C<ifndef> must be on a line by itself.
+
+Hiding the declaration completely is done when OpenSSL is configured to
+remove old API's, via the C<--api> configuration parameter.
```
[the remark this answers is not in the document]
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:101]: fixed, thanks.

**Comment by levitte** [comment]
Note that as response to this, I did a bunch of edits, see #13240

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 4 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:113]: I disagree:
- I think most devs will look to existing code, especially header files
- Picking the right words for the section title is *hard,* as we've seen
- Dividing it up into sections breaks the narrative flow.

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 7 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:122]: So what's the difference between the header file and the simple macro? I thought I understood it, but I guess I don't.

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 8 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:129]: I disagree about a section title, for the same reasons as above https://github.com/openssl/openssl/pull/13198#discussion_r512149471

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 9 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:132]: Then what is the point of the header file?  And when should it be used?

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 11 above)
Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:139]: Disagree about section titles, for reasons stated in https://github.com/openssl/openssl/pull/13198#discussion_r512149471

File: doc/internal/man7/OSSL_DEPRECATED.pod (hunk 4 above)
Comment (levitte) [reply doc/internal/man7/OSSL_DEPRECATED.pod:113]: It's not immediately apparent which code section goes there, especially the two variants that go into translation units.

Comment (richsalz) [reply doc/internal/man7/OSSL_DEPRECATED.pod:113]: I think we disagree, but I am not totally sure that I understand your comment.  "translation units" might be pedantically correct, but OpenSSL has files. :)

**Comment by richsalz** [comment]
I did yet another rewrite; there are subsection headings now.  I still don't understand the difference between the header file and the macro, as in https://github.com/openssl/openssl/pull/13198#discussion_r511544153

**Pull request closed without merging**